{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"FINAL version of calculating Splicing Efficiency.\n",
    "\n",
    "\n",
    "JunctionSeq gives weird SJ coordinates: as start it sets last exonic base, and as end it sets last intronic base.\n",
    "Whereas STAR aligner gives coordinates for intronic bases only.\n",
    "Example for YGL030W/RPL30 gene: \n",
    "    STAR SJ: VII:439094-439323\n",
    "    JunctionSeq SJ: VII:439093-439323  \n",
    "\n",
    "This method uses coordinates generated by JunctionSeq. \n",
    "\n",
    "To get read counts we use coordinates for the last base of exon_1 (to extract exon-intron read counts at 5'SS) \n",
    "and first base of exon_2 (to extract intron-exon read counts at 3'SS)\"\"\"\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "def exe(cmd):\n",
    "    import subprocess\n",
    "    o = subprocess.Popen(\n",
    "        cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
    "    out = o.stdout.read().strip().decode()\n",
    "    err = o.stderr.read().strip().decode()\n",
    "    return out, err\n",
    "\n",
    "def spleff(decoder, samplist, bampath, outfolder, verbose=False): \n",
    "\n",
    "    \"\"\"Calculation of splicing effciency.\n",
    "    \n",
    "    --> This method takes into account that when exon_1 is very short (<30 nt), 'p' value equals 0, \n",
    "        and so 'total' = EE + EI3.\n",
    "        For novel junctions, which are upstream start of exon_1, intron length is gonna get negative value.\n",
    "    --> Importantly, gene orientation is taken into account too!     \n",
    "    \n",
    "    Attributes:\n",
    "\n",
    "        decoder (string) : path to a file generated by JunctionSeq; this file name ends with \"bySampleallGenes.results.txt\";\n",
    "            particularly you need columns named: 'featureID', 'geneID', 'chr', 'start', 'end', \n",
    "            and 'CDSstart', 'CDSend', 'Strand' (these you need to add to a table by yourself BEFORE running this program! save it as excel file);\n",
    "            'Strand' is needed to account for gene orientation (which values are junction 'start' and which are 'stop')\n",
    "            featureID needs to look like this: YGL030W:J003\n",
    "        \n",
    "        samplist (list) : list of sample names, e.g. \"MEX_WT_7\", \"MEX_WT_8\", \"MEX_MUT_10\", \"MEX_MUT_11\";\n",
    "            names in this list are simply a part of each BAM file name, e.g. from file name \"MEX_WT_7Aligned.sortedByCoord.out.bam\" use \"MEX_WT_7\"\n",
    "        \n",
    "        bampath (string): part of the path to each BAM file shared between all BAM files to be used, \n",
    "            e.g. \"/Volumes/agata/Desktop/RNAseq/output_copied/[...]\"\n",
    "            Important for other purposes --> BAM file name has to end with: \"*Aligned.sortedByCoord.out.bam\"\n",
    "        \n",
    "        outfolder (string): path to a folder where you store resulting excel files  \n",
    "        \n",
    "    Example::\n",
    "    \n",
    "         decoderMex = \"/Users/agata/Desktop/RNAseq_results/opt/QoRTs_JunctionSeq/2nd_approach/Splicing_Efficiency/decoder_MEX67_exStartStopStrand.xls\"\n",
    "         samplistMex = [\"MEX_WT_7\", \"MEX_WT_8\", \"MEX_WT_9\", \"MEX_MUT_10\", \"MEX_MUT_11\", \"MEX_MUT_12\"]\n",
    "         bampath = \"/Volumes/agata/Desktop/RNAseq/output_copied/\"\n",
    "         outfolder = \"/Users/agata/Desktop/RNAseq_results/opt/QoRTs_JunctionSeq/2nd_approach/Splicing_Efficiency/\"\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    df = pd.read_excel(decoder)\n",
    "    for sample in samplist:  \n",
    "        dfout = pd.DataFrame()\n",
    "        bam = bampath + sample + \"Aligned.sortedByCoord.out.bam\"\n",
    "        counter = 0\n",
    "        for index, row in df.iterrows():\n",
    "            f = row['featureID']\n",
    "            gene = row['geneID']\n",
    "            ch = row['chr']\n",
    "            start = row['start']\n",
    "            end = row['end']\n",
    "            CDSstart = row['CDSstart']\n",
    "            CDSend = row['CDSend']\n",
    "            strand = row['Strand']\n",
    "            gene_name, element = f.split(':')\n",
    "            if element.startswith('J') or element.startswith('N'):\n",
    "                if strand == \"+\":\n",
    "                    midintron = start + int((end - start)/2)\n",
    "                    cmd = \"/usr/local/bin/samtools view \" + bam + \" \" + ch + \":\" + str(midintron) + \"-\" + str(midintron) + \"| cut -f 6 > /tmp/SE\" \n",
    "                    out, err = exe(cmd)\n",
    "                    cmd = \"cat /tmp/SE | grep 'N' | wc -l\"\n",
    "                    EE_ct, err = exe(cmd)\n",
    "                    EE_ct = int(EE_ct)\n",
    "\n",
    "                    cmd = \"/usr/local/bin/samtools view \" + bam + \" \" + ch + \":\" + str(start) + \"-\" + str(start) + \"| cut -f 6 > /tmp/SE\" \n",
    "                    out, err = exe(cmd)\n",
    "                    cmd = \"cat /tmp/SE | grep -v 'N' | wc -l\"\n",
    "                    EI5_ct, err = exe(cmd)\n",
    "                    EI5_ct = int(EI5_ct)\n",
    "\n",
    "                    cmd = \"/usr/local/bin/samtools view \" + bam + \" \" + ch + \":\" + str(int(end)+1) + \"-\" + str(int(end)+1) + \"| cut -f 6 > /tmp/SE\" \n",
    "                    out, err = exe(cmd)\n",
    "                    cmd = \"cat /tmp/SE | grep -v 'N' | wc -l\"\n",
    "                    EI3_ct, err = exe(cmd)\n",
    "                    EI3_ct = int(EI3_ct)\n",
    "                \n",
    "                    ex1_len = start - CDSstart\n",
    "                    if ex1_len < 30:\n",
    "                        p = 0\n",
    "                        total = EE_ct + EI3_ct\n",
    "                        se = round(EE_ct / total, 4)\n",
    "                    else:    \n",
    "                        if (EI5_ct + EI3_ct) != 0 and EE_ct != 0:\n",
    "                            p = round(EI5_ct / (EI5_ct + EI3_ct), 6)\n",
    "                            total = round(EE_ct + (p * EI5_ct) + (1 - p) * EI3_ct, 2)\n",
    "                            se = round(EE_ct / total, 4)\n",
    "                        elif (EI5_ct + EI3_ct) == 0 and EE_ct != 0:\n",
    "                            p = 0\n",
    "                            total = EE_ct\n",
    "                            se = 1\n",
    "                        elif (EI5_ct + EI3_ct) != 0 and EE_ct == 0:\n",
    "                            p = round(EI5_ct / (EI5_ct + EI3_ct), 6)\n",
    "                            total = round(EE_ct + (p * EI5_ct) + (1 - p) * EI3_ct, 2)\n",
    "                            se = 0\n",
    "                        else:\n",
    "                            p = \"N/A\"\n",
    "                            total = \"N/A\"\n",
    "                            se = \"N/A\"\n",
    "                if strand == \"-\":\n",
    "                    midintron = start + int((end - start)/2)\n",
    "                    cmd = \"/usr/local/bin/samtools view \" + bam + \" \" + ch + \":\" + str(midintron) + \"-\" + str(midintron) + \"| cut -f 6 > /tmp/SE\" \n",
    "                    out, err = exe(cmd)\n",
    "                    cmd = \"cat /tmp/SE | grep 'N' | wc -l\"\n",
    "                    EE_ct, err = exe(cmd)\n",
    "                    EE_ct = int(EE_ct)\n",
    "\n",
    "                    cmd = \"/usr/local/bin/samtools view \" + bam + \" \" + ch + \":\" + str(start) + \"-\" + str(start) + \"| cut -f 6 > /tmp/SE\" \n",
    "                    out, err = exe(cmd)\n",
    "                    cmd = \"cat /tmp/SE | grep -v 'N' | wc -l\"\n",
    "                    EI3_ct, err = exe(cmd)\n",
    "                    EI3_ct = int(EI3_ct)\n",
    "\n",
    "                    cmd = \"/usr/local/bin/samtools view \" + bam + \" \" + ch + \":\" + str(int(end)+1) + \"-\" + str(int(end)+1) + \"| cut -f 6 > /tmp/SE\" \n",
    "                    out, err = exe(cmd)\n",
    "                    cmd = \"cat /tmp/SE | grep -v 'N' | wc -l\"\n",
    "                    EI5_ct, err = exe(cmd)\n",
    "                    EI5_ct = int(EI5_ct)\n",
    "                    \n",
    "                    ex1_len = CDSend - end\n",
    "                    if ex1_len < 30:\n",
    "                        p = 0\n",
    "                        total = EE_ct + EI3_ct\n",
    "                        se = round(EE_ct / total, 4)\n",
    "                    else:    \n",
    "                        if (EI5_ct + EI3_ct) != 0 and EE_ct != 0:\n",
    "                            p = round(EI5_ct / (EI5_ct + EI3_ct), 6)\n",
    "                            total = round(EE_ct + (p * EI5_ct) + (1 - p) * EI3_ct, 2)\n",
    "                            se = round(EE_ct / total, 4)\n",
    "                        elif (EI5_ct + EI3_ct) == 0 and EE_ct != 0:\n",
    "                            p = 0\n",
    "                            total = EE_ct\n",
    "                            se = 1\n",
    "                        elif (EI5_ct + EI3_ct) != 0 and EE_ct == 0:\n",
    "                            p = round(EI5_ct / (EI5_ct + EI3_ct), 6)\n",
    "                            total = round(EE_ct + (p * EI5_ct) + (1 - p) * EI3_ct, 2)\n",
    "                            se = 0\n",
    "                        else:\n",
    "                            p = \"N/A\"\n",
    "                            total = \"N/A\"\n",
    "                            se = \"N/A\"\n",
    "\n",
    "                counter = counter + 1\n",
    "                \n",
    "                print(gene, f, ch, start, end, EI5_ct, EI3_ct, EE_ct, p, total, ex1_len, se)\n",
    "\n",
    "                dfout = dfout.append(pd.DataFrame({'geneID': gene,\n",
    "                                                   'chr': ch,\n",
    "                                                   'strand' : strand,\n",
    "                                                   'featureID': f,\n",
    "                                                   'I_5p': start,\n",
    "                                                   'I_3p': end,\n",
    "                                                   'ex1_length_' + sample : ex1_len,\n",
    "                                                   'EI5_rawCt_' + sample : EI5_ct,\n",
    "                                                   'EI3_rawCt_' + sample : EI3_ct,\n",
    "                                                   'EE_rawCt_' + sample : EE_ct,\n",
    "                                                   'p_' + sample : p,\n",
    "                                                   'total_' + sample : total,\n",
    "                                                   'SE_' + sample : se,\n",
    "                                                  }, index=[counter]))\n",
    "        \n",
    "        \n",
    "        outpath = outfolder + \"SE_\" + sample + '_p.And.ExLen.xls'\n",
    "        if verbose: \n",
    "            print('outpath:', outpath)\n",
    "        dfout.to_excel(outpath)\n",
    "\n",
    "\n",
    "    \n",
    "decoderMex = \"/Users/agata/Desktop/RNAseq_results/opt/QoRTs_JunctionSeq/2nd_approach/Splicing_Efficiency/decoder_MEX67_exStartStopStrand.xls\"\n",
    "samplistMex = [\"MEX_WT_7\", \"MEX_WT_8\", \"MEX_WT_9\", \"MEX_MUT_10\", \"MEX_MUT_11\", \"MEX_MUT_12\"]\n",
    "\n",
    "decoderNpl3 = \"/Users/agata/Desktop/RNAseq_results/opt/QoRTs_JunctionSeq/2nd_approach/Splicing_Efficiency/decoder_NPL3_exStartStopStrand.xls\"\n",
    "samplistNpl3 = [\"N3_WT_1\", \"N3_WT_2\", \"N3_WT_3\", \"DL_4\", \"DL_5\", \"DL_6\"]\n",
    "\n",
    "decoder305 = \"/Users/agata/Desktop/RNAseq_results/opt/QoRTs_JunctionSeq/2nd_approach/Splicing_Efficiency/decoder_305_exStartStopStrand.xls\"\n",
    "samplist305 = [\"MTR10_WT_13\", \"MTR10_WT_14\", \"MTR10_WT_15\", \"305_16\", \"305_17\", \"305_18\"]\n",
    "\n",
    "decoder406 = \"/Users/agata/Desktop/RNAseq_results/opt/QoRTs_JunctionSeq/2nd_approach/Splicing_Efficiency/decoder_406_exStartStopStrand.xls\"\n",
    "samplist406 = [\"406_19\", \"406_20\", \"406_21\"]\n",
    "\n",
    "bampath = \"/Volumes/agata/Desktop/RNAseq/output_copied/\"\n",
    "\n",
    "outfolder = '/Users/agata/Desktop/RNAseq_results/opt/QoRTs_JunctionSeq/2nd_approach/Splicing_Efficiency/SE_p-calc/p-calc_ex1len/'\n",
    "\n",
    "spleff(decoderMex, samplistMex, bampath, outfolder)\n",
    "spleff(decoderNpl3, samplistNpl3, bampath, outfolder)\n",
    "spleff(decoder305, samplist305, bampath, outfolder)\n",
    "spleff(decoder406, samplist406, bampath, outfolder)\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
